sound/soc/soc-pcm.c

```diff
@@ -2744,29 +2744,19 @@ int soc_new_pcm(struct snd_soc_pcm_runtime *rtd, int num)
 			stream = SNDRV_PCM_STREAM_PLAYBACK;
 
 			for_each_rtd_cpu_dais(rtd, i, cpu_dai)
-				if (!snd_soc_dai_stream_valid(cpu_dai,
-							      stream)) {
-					dev_err(rtd->card->dev,
-						"CPU DAI %s for rtd %s does not support playback\n",
-						cpu_dai->name,
-						rtd->dai_link->stream_name);
-					return -EINVAL;
+				if (snd_soc_dai_stream_valid(cpu_dai, stream)) {
+					playback = 1;
+					break;
 				}
-			playback = 1;
 		}
 		if (rtd->dai_link->dpcm_capture) {
 			stream = SNDRV_PCM_STREAM_CAPTURE;
 
 			for_each_rtd_cpu_dais(rtd, i, cpu_dai)
-				if (!snd_soc_dai_stream_valid(cpu_dai,
-							      stream)) {
-					dev_err(rtd->card->dev,
-						"CPU DAI %s for rtd %s does not support capture\n",
-						cpu_dai->name,
-						rtd->dai_link->stream_name);
-					return -EINVAL;
+				if (snd_soc_dai_stream_valid(cpu_dai, stream)) {
+					capture = 1;
+					break;
 				}
-			capture = 1;
```

Comment:
Hi @plbossart. A bit weird that you post these patches on GitHub first before sending them to the mailing list :)

Anyway, I agree with this part of the patch. Essentially this is some derivation of what I suggested here: https://lore.kernel.org/alsa-devel/20200617174635.GA40060@gerhold.net/

You no longer error out if `dpcm_playback`/`dpcm_capture` isn't set exactly like the capabilities of the DAIs, so this restores the original behavior (before your `ASoC: soc-pcm: dpcm: fix playback/capture checks`). `dpcm_playback`/`dpcm_capture` only act as an additional restriction of the capabilities of a DAI link, but there is no need to set them correctly if you just want to expose everything that is possible with the underlying DAIs.

Doesn't that make the `snd_soc_dai_link_set_capabilities` unnecessary again? You could just remove it and restore `dpcm_playback = dpcm_capture = 1` for simple-audio-card.

Reply:
@Minecrell we use GitHub to trigger our CI tests. We always post here before sending to the mailing list.

You're right that the helper is not useful with this code, but we still have a case where we can have invalid configurations. I will re-add a check so that we return an invalid config when the dailink dpcm_playback =1 but none of the dais support playback, same for capture.

Reply:
@Minecrell see updated code

Reply:
@plbossart It looks fine, although personally I think the validation isn't that useful. IMO the DAI capabilities should always determine the possible directions for a DAI link. Requiring everyone to set `dpcm_playback` and `dpcm_capture` correctly is just redundant information that can be easily inferred from the given DAIs.

Plus, I'm not sure that there are that many use cases for artificially limiting the DAI direction (if any). And even if there is some rare use case, then we could just have some "restriction parameters" that will override the default automatic detection.

But I think I have mentioned this several times now here and in the other mail thread. At the end I'm fine with your solution as well as long as it works :)

Reply:
The DAIs may be exposed by different components, and I think it does help for the integration to verify there's no consistency issues.

We do have HDMI, DMIC, SoundWire DAIs that are playback or capture only. It's really for those cases that the tests made sense. When a dai is full-duplex yes all of this is of limited value.